src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/formula/JoinAwareWeightedUnboundVariableCount.java

```diff
@@ -0,0 +1,68 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.formula;
+
+import org.apache.jena.graph.Node;
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.Join_Analyzer;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.SubQuery_Analyzer;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.Var_Analyzer;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class JoinAwareWeightedUnboundVariableCount {
+    public static final double W_S = 1; // weight for subject variables
+    public static final double W_P = 0.1; // weight for predicate variables
+    public static final double W_O = 0.8; // weight for object variables
+
+    public static final double J_Tc=0.6; // weight for chain join
+    public static final double J_Ts=0.5; // weight for star join
+    public static final double J_Tu=1; // weight for unusual join
+    public static final double J_Tn=0; // no join
+
+    public static double estimate( final List<LogicalPlan> selectedPlans, LogicalPlan lop ) {
+        Set<Node> bindings = new HashSet<>();
+        for ( LogicalPlan plan : selectedPlans ) {
+            final SubQuery_Analyzer analyzer = new SubQuery_Analyzer( plan );
+            bindings = Var_Analyzer.addBinds( analyzer.getSubs(), analyzer.getPreds(), analyzer.getObjs() );
+        }
```

Comment:
For points 3 and 4, I have added 'plan' as a member of 'Query_Analyzer', and used 'Query_Analyzer' as elements of 'candidatePlans' and 'selectedSubPlans' to avoid repeatedly creating Query_Analyzer for the same sub-plan. 
(see commit [22b0b7b](https://github.com/LiUSemWeb/HeFQUIN/pull/277/commits/22b0b7bb2bdbced295bb2bbbe3fb2d04de01bf66))